fix: backward compatibility with version 7.17.0

salvatore-campagna opened the pull request:
PR #83339 was meant to be included in version 7.17.0 but didn't happen.
It landed in version 7.17.1. As a result, version checks in the code are
incorrect. Instead of checking against version 7.17.0 we need to adjust
the code in such a way that it operates correctly with version 7.17.0.
Version 7.17.0 is released and, as as result, it represents the 'reference'
version we would like all following versions to interoperate with.

Unfortunately this means that interoperability between version 7.17.0 and
version 8.0.0 is broken.

In summary this fix needs to be backported to:
* 7.17.1: so to fix interoperability of version 7.17.1 with 7.17.0
* 8.0.1: so to fix interoperability of version 8.0.1 with 7.17.0
* 8.1 (currently BC): so to fix interoperability of version 8.1.0 with 7.17.0
* master (8.2): so to fix interoperability of version 8.2.0 with 7.17.0

Also, when reading and writing the range key, version 7.17.0 checks against
version 6.4.0 which is not possible with the current version in master.

This is a fix for #83533.

salvatore-campagna committed:
fix: backward compatibility with version 7.17.0

PR #83339 was meant to be included in version 7.17.0 but didn't happen.
It landed in version 7.17.1. As a result, version checks in the code are
incorrect. Instead of checking against version 7.17.0 we need to adjust
the code in such a way that it operates correctly with version 7.17.0.
Version 7.17.0 is released and, as as result, it represents the 'reference'
version we would like all following versions to interoperate with.

Unfortunately this means that interoperability between version 7.17.0 and
version 8.0.0 is broken.

In summary this fix needs to be backported to:
* 7.17.1: so to fix interoperability of version 7.17.1 with 7.17.0
* 8.0.1: so to fix interoperability of version 8.0.1 with 7.17.0
* 8.1 (currently BC): so to fix interoperability of version 8.1.0 with 7.17.0
* master (8.2): so to fix interoperability of version 8.2.0 with 7.17.0

Also, when reading and writing the range key, version 7.17.0 checks against
version 6.4.0 which is not possible right now. As a result, we just check
against the earliest possible version.

salvatore-campagna committed:
fix: update skip version to include the revision

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

elasticsearchmachine commented:
Hi @salvatore-campagna, I've created a changelog YAML for you.

salvatore-campagna committed:
Update docs/changelog/83715.yaml

salvatore-campagna commented:
@elasticmachine run elasticsearch-ci/part-2

salvatore-campagna committed:
fix: always generate the key before writing

salvatore-campagna committed:
fix: key generation for versions above 17.0.1

nik9000 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalRange.java:
```diff
@@ -157,6 +157,7 @@ private static String generateKey(double from, double to, DocValueFormat format)
 
         @Override
         public void writeTo(StreamOutput out) throws IOException {
+            // NOTE: we would actually need to use Version.V_6_4_0 which is not available, so we use the earliest possible version
```
I don't think 7_17_1 is the earliest possible version though, right?

salvatore-campagna replied:
Yes I will remove the comment...I had to change that in one of the following commits...my assumption about the earlier version was incorrect.

nik9000 commented:
I think the version changes look right to me. Do you have to change and BWC test skips in this one too?

salvatore-campagna committed:
fix: remove incorrect comment

salvatore-campagna commented:
>

salvatore-campagna commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix/7.17.0-bwc-compatibility

salvatore-campagna commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix/7.17.0-bwc-compatibility

salvatore-campagna commented:
@elasticmachine run elasticsearch-ci/test-full-bwc

salvatore-campagna commented:
@elasticmachine run elasticsearch-ci/bwc

salvatore-campagna commented:
@elasticmachine test this

salvatore-campagna commented:
@elasticmachine test this please

salvatore-campagna commented:
@elasticmachine run elasticsearch-ci/part-2

salvatore-campagna commented:
I think this is still not enough. Reading the optional double values `originalFrom` and `originalTo` needs to be done only if version is exactly `7.17.0`. `7.17.0` is the only version writing `originalFrom` and `originalTo`.

salvatore-campagna committed:
fix: limit usage of 'originalFrom' and 'originalTo' to version 7.17.0

Here we also re-enable test "40_range/Float test" which was fixed in
version 7.17.1.

salvatore-campagna commented:
I have executed backward compatibility tests locally against version `7.17.0`, `7.17.1`, `8.0.0` and `8.1.0` multiple times...I took advantage of randomness to run something like 4 or 5 backward compatibility tests against each version. I didn't see any failure. Also test "Float range" which covers the original issue works fine.

salvatore-campagna committed:
fix: code format violations

salvatore-campagna committed:
fix: handle null values as positive and negative infinity

salvatore-campagna commented:
@elasticmachine run elasticsearch-ci/full-bwc

elasticmachine committed:
Merge branch 'master' into fix/7.17.0-bwc-compatibility

salvatore-campagna committed:
fix: drop the upper bound on version 7.17.1

Versions 7.17.1 and all the following versions just check to be
on or after version 7.17.0 to decide whether reading/writing the
originalFrom and originalTo values. As a result, this new version
needs to work with those versions too.

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.1](https://github.com/elastic/elasticsearch/pull/83819)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/83819">](https://github.com/elastic/elasticsearch/pull/83819) |
| ✅ |  [7.17](https://github.com/elastic/elasticsearch/pull/83820)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/83820">](https://github.com/elastic/elasticsearch/pull/83820) |
| ✅ |  [8.0](https://github.com/elastic/elasticsearch/pull/83821)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/83821">](https://github.com/elastic/elasticsearch/pull/83821) |
